### PR TITLE
#2020 - Remediate high risk vulnerabilities from WAVA report (API Mass Assignment)

### PR DESCRIPTION
**Pull request by andrewsignori-aot**
### Context

Enabled the Nestjs configuration `forbidNonWhitelisted` to force the API to throw an error and return a "bad request" if some property not mapped to the DTO is received (more info on [forbidNonWhitelisted](https://docs.nestjs.com/techniques/validation)).

Currently, the API already has the configuration `whitelist` which will remove any DTO-non-mapped-property from the payload received. Enabling the `forbidNonWhitelisted`  will allow the API to return a better response and also restrict unknown query string parameters.

#### PR Goal

The PR intention is primarily to enable the API configuration and adjust the HTTP call from the web application to ensure that non-extra properties will be sent. The web application sends extra values when the payload is not "manually created" which happens basically when the form.io submissions objects are sent directly to the API. These payloads contain the properties needed by the API and enforced by the DTO but they also contain some extra non-expected values that were part of the form.io form but do not need to be part of the HTTP request.

To enforce the DTO properties, the class-transformer package was used (the same used by Nestjs) to convert a "plain object", like the form.io submission, to a DTO class, also removing any extra property along the process.

Not every form.io definition had its output converted to a class because not every form.io form is consumed in the same way.
Please see below the ones affected.

Typed to a class to exclude extraneous values.

- educationprogram
- educationprogramoffering
- institutionlocation
- institutionprofile
- institutionprofilecreation
- institutionuserprofile
- programinformationrequest
- uploadstudentdocuments
- studentprofile
- studentexceptions

No changes are needed because the payload is controlled.

- confirmsstudentenrollment
- designationagreementdetails
- reportscholasticstandingchange
- uploadstudentdocumentsaest
- exportfinancialreports
- staffapprovalappeal
- studentrequestchange

To be removed in an upcoming PR.

- approvedeclineoffering
- approvedenydesignation
- approveeducationprogram
- createnote
- declineeducationprogram
- studentapplicationdetails
- trackstudentapplication

### CODE REVIEW

File: sources/packages/backend/apps/api/src/utilities/auth-utils.ts [hunk 1]
```diff
@@ -46,6 +46,7 @@ export function setGlobalPipes(app: INestApplication) {
   app.useGlobalPipes(
     new ValidationPipe({
       whitelist: true,
+      forbidNonWhitelisted: true,
```
Comment (dheepak-aot) [inline sources/packages/backend/apps/api/src/utilities/auth-utils.ts:49]: 👍

File: sources/packages/backend/apps/api/src/route-controllers/institution-locations/models/application.dto.ts [hunk 2]
```diff
@@ -44,6 +47,15 @@ export class ActiveApplicationSummaryAPIOutDTO {
   applicationScholasticStandingStatus: ApplicationScholasticStandingStatus;
 }
 
+/**
+ * Query string options available for active applications summary.
+ */
+export class ActiveApplicationSummaryAPIQueryStringDTO extends ApplicationStatusPaginationOptionsAPIInDTO {
+  @IsBoolean()
+  @ToBoolean()
```
Comment (dheepak-aot) [inline sources/packages/backend/apps/api/src/route-controllers/institution-locations/models/application.dto.ts:55]: Trying to understand why we are using `@ToBoolean` here. 
Params like page and pageLimit are casted as number in the DTO.

File: sources/packages/backend/apps/api/src/route-controllers/institution-locations/institution-location.institutions.controller.ts [hunk 3]
```diff
@@ -155,14 +151,12 @@ export class InstitutionLocationInstitutionsController extends BaseController {
...
-    @Query() pagination: ApplicationStatusPaginationOptionsAPIInDTO,
-    @Query("archived", new DefaultValuePipe(false), ParseBoolPipe)
-    archived: boolean,
+    @Query() queryStringDTO: ActiveApplicationSummaryAPIQueryStringDTO,
   ): Promise<PaginatedResultsAPIOutDTO<ActiveApplicationSummaryAPIOutDTO>> {
     const applications = await this.applicationService.getActiveApplications(
       locationId,
-      pagination,
-      archived,
+      queryStringDTO,
```
Comment (dheepak-aot) [inline sources/packages/backend/apps/api/src/route-controllers/institution-locations/institution-location.institutions.controller.ts:158]: Please update the comment with modified param.

Also param name queryStringDTO seems to be very generic.

File: sources/packages/backend/apps/api/src/services/education-program/education-program.service.models.ts [hunk 4]
```diff
@@ -1,14 +1,19 @@
 import { ProgramIntensity, ProgramStatus } from "@sims/sims-db";
+import { Allow } from "class-validator";
 
 export interface ProgramDeliveryTypes {
   deliveredOnSite: boolean;
   deliveredOnline: boolean;
 }
 
-export interface EntranceRequirements {
+export class EntranceRequirements {
```
Comment (dheepak-aot) [inline sources/packages/backend/apps/api/src/services/education-program/education-program.service.models.ts:9]: May be or may not be part of the PR. EntranceRequirements is a service model which by mistake we are using in the DTO level in EducationProgramAPIOutDTO. We can create a class for EntranceRequirements  at DTO level and add allow decorator to it.

File: sources/packages/backend/apps/api/src/route-controllers/education-program/_tests_/e2e/education-program.institutions.controller.createEducationProgram.e2e-spec.ts [hunk 5]
```diff
@@ -83,17 +84,16 @@ describe("EducationProgramInstitutionsController(e2e)-createEducationProgram", (
...
       testingModule,
       AppInstitutionsModule,
       FormService,
     );
-    formService.dryRunSubmission = jest
-      .fn()
-      .mockResolvedValue({ valid: true, data: { data: payload } });
+    formService.dryRunSubmission = jest.fn().mockResolvedValue({
+      valid: true,
+      data: { data: { ...payload, programStatus } },
```
Comment (ann-aot) [inline sources/packages/backend/apps/api/src/route-controllers/education-program/_tests_/e2e/education-program.institutions.controller.createEducationProgram.e2e-spec.ts:95]: 👍

File: sources/packages/backend/apps/api/src/route-controllers/education-program/_tests_/e2e/education-program.institutions.controller.createEducationProgram.e2e-spec.ts [hunk 6]
```diff
@@ -107,6 +107,7 @@ describe("EducationProgramInstitutionsController(e2e)-createEducationProgram", (
       .auth(institutionUserToken, BEARER_AUTH_TYPE)
       .expect(HttpStatus.CREATED)
       .then((response) => {
+        console.log(response.body);
```
Comment (ann-aot) [inline sources/packages/backend/apps/api/src/route-controllers/education-program/_tests_/e2e/education-program.institutions.controller.createEducationProgram.e2e-spec.ts:110]: pls remove the console.log

File: sources/packages/backend/apps/api/src/route-controllers/institution-locations/institution-location.institutions.controller.ts [hunk 7]
```diff
@@ -155,14 +151,12 @@ export class InstitutionLocationInstitutionsController extends BaseController {
...
+      queryStringDTO.archived,
```
Comment (ann-aot) [inline sources/packages/backend/apps/api/src/route-controllers/institution-locations/institution-location.institutions.controller.ts:159]: 👍

File: sources/packages/backend/apps/api/src/route-controllers/institution-locations/models/application.dto.ts [hunk 8]
```diff
@@ -44,6 +47,15 @@ export class ActiveApplicationSummaryAPIOutDTO {
   applicationScholasticStandingStatus: ApplicationScholasticStandingStatus;
 }
 
+/**
+ * Query string options available for active applications summary.
+ */
+export class ActiveApplicationSummaryAPIQueryStringDTO extends ApplicationStatusPaginationOptionsAPIInDTO {
```
Comment (ann-aot) [inline sources/packages/backend/apps/api/src/route-controllers/institution-locations/models/application.dto.ts:53]: just to double are we not using `APIInDTO` for query string STO?

File: sources/packages/web/src/components/common/OfferingFormSubmit.vue [hunk 9]
```diff
@@ -114,11 +112,15 @@ export default defineComponent({
     const validateOfferingData = async (
       data: EducationProgramOfferingAPIInDTO,
     ): Promise<OfferingValidationResultAPIOutDTO> => {
+      const typedData = excludeExtraneousValues(
```
Comment (ann-aot) [inline sources/packages/web/src/components/common/OfferingFormSubmit.vue:115]: 👍

File: sources/packages/web/src/services/InstitutionService.ts [hunk 10]
```diff
@@ -52,7 +53,7 @@ export class InstitutionService {
   }
 
   async updateInstitution(
-    data: InstitutionContactAPIInDTO,
```
Comment (dheepak-aot) [inline sources/packages/web/src/services/InstitutionService.ts:55]: Optional. We can add comments to the method.

File: sources/packages/web/src/services/http/dto/Common.dto.ts [hunk 11]
```diff
@@ -16,16 +18,26 @@ export interface InstitutionPrimaryContactAPIOutDTO {
  * a generic address interface that is used extend or used as nested type
  * for both IN and OUT DTOs
  */
-export interface AddressDetailsFormAPIDTO {
+export class AddressDetailsFormAPIDTO {
```
Comment (dheepak-aot) [inline sources/packages/web/src/services/http/dto/Common.dto.ts:21]: APIInDTO?

File: sources/packages/web/src/services/http/dto/ScholasticStanding.dto.ts [hunk 12]
```diff
@@ -5,7 +5,7 @@ export const INVALID_OPERATION_IN_THE_CURRENT_STATUS =
   "INVALID_OPERATION_IN_THE_CURRENT_STATUS";
 
 export interface ScholasticStandingDataAPIInDTO {
```
Comment (dheepak-aot) [inline sources/packages/web/src/services/http/dto/ScholasticStanding.dto.ts:7]: This DTO is still an interface?

Comment (dheepak-aot) [reply sources/packages/web/src/services/http/dto/ScholasticStanding.dto.ts:7]: There is one more DTO which extends this and its an interface. 
image

File: sources/packages/web/src/services/DesignationAgreementService.ts [hunk 13]
```diff
@@ -55,19 +57,23 @@ export class DesignationAgreementService {
...
       );
+    const { excludeExtraneousValues } = useFormioUtils();
+    const typedPayload = excludeExtraneousValues(
+      UpdateDesignationDetailsAPIInDTO,
+      designation,
+    );
     await ApiClient.DesignationAgreement.updateDesignationAgreement(
       designationId,
-      designation,
+      typedPayload,
```
Comment (ann-aot) [inline sources/packages/web/src/services/DesignationAgreementService.ts:76]: In my local, my getting below error, 
image

File: sources/packages/web/src/services/http/dto/Institution.dto.ts [hunk 14]
```diff
@@ -90,24 +104,54 @@ export interface InstitutionBasicAPIOutDTO {
...
+  institutionType: number;
+  mailingAddress: AddressDetailsFormAPIDTO;
+}
+
 /**
  * DTO for institution creation by the institution user during the on board process
  * when the institution profile and the admin user must be created altogether.
  */
-export interface CreateInstitutionAPIInDTO {
+export class CreateInstitutionAPIInDTO implements CreateInstitution {
```
Comment (dheepak-aot) [inline sources/packages/web/src/services/http/dto/Institution.dto.ts:127]: Just to clarify, will the properties of CreateInstitution be considered by plainToClass as they are not decorated?

File: sources/packages/web/src/components/common/OfferingFormSubmit.vue [hunk 15]
```diff
@@ -114,11 +112,15 @@ export default defineComponent({
...
+        EducationProgramOfferingAPIInDTO,
+        data,
+      );
       const validationResult =
         await EducationProgramOfferingService.shared.validateOffering(
           props.locationId,
           props.programId,
-          data,
+          typedData,
```
Comment (ann-aot) [inline sources/packages/web/src/components/common/OfferingFormSubmit.vue:123]: In my local, Iam getting error for typedData
image

File: sources/packages/web/src/components/partial-view/DesignationAgreement/DesignationAgreementForm.models.ts [hunk 16]
```diff
@@ -46,3 +46,29 @@ export interface DesignationModel {
   designationStatus?: DesignationAgreementStatus;
   designationStatusClass?: string;
 }
+
+/**
+ * Details for approve or update an approved designation agreement.
+ */
+export interface UpdateDesignationDetailsModel {
+  designationStatus: DesignationAgreementStatus;
+  startDate?: string;
```
Comment (ann-aot) [inline sources/packages/web/src/components/partial-view/DesignationAgreement/DesignationAgreementForm.models.ts:55]: Just  to double-check, won't startDate and endDate are required during the approval process?

File: sources/packages/web/src/components/common/OfferingFormSubmit.vue (hunk 15 above)
Comment (ann-aot) [reply sources/packages/web/src/components/common/OfferingFormSubmit.vue:123]: never mind, it was throwing the error, because, I didn't install `class-transformer` on the web. resolving

File: sources/packages/web/src/views/aest/institution/ProfileCreate.vue [hunk 17]
```diff
@@ -36,17 +36,22 @@ export default defineComponent({
...
     const snackBar = useSnackBar();
+    const { excludeExtraneousValues } = useFormioUtils();
     const processing = ref(false);
     const institutionProfileModel = ref({
       clientType: ClientIdType.AEST,
       mode: "create",
     } as FormModel);
 
-    const createInstitution = async (data: AESTCreateInstitutionAPIInDTO) => {
+    const createInstitution = async (data: unknown) => {
```
Comment (ann-aot) [inline sources/packages/web/src/views/aest/institution/ProfileCreate.vue:46]: why it was updated to `unknown`

File: sources/packages/web/src/views/aest/institution/InstitutionLocationEdit.vue [hunk 18]
```diff
@@ -36,18 +36,21 @@ export default defineComponent({
   setup(props) {
     const initialData = ref({} as InstitutionLocationEdit);
     const snackBar = useSnackBar();
+    const { excludeExtraneousValues } = useFormioUtils();
     const router = useRouter();
-    const updateInstitutionLocation = async (
-      data: InstitutionLocationAPIInDTO,
-    ) => {
+    const updateInstitutionLocation = async (data: unknown) => {
```
Comment (dheepak-aot) [inline sources/packages/web/src/views/aest/institution/InstitutionLocationEdit.vue:41]: Should we declare unknown here where the submit/emit happens?

image

**Review by ann-aot — COMMENTED** [review]
Good work @andrewsignori-aot 👍 Added some comments

File: sources/packages/web/src/views/aest/institution/ProfileCreate.vue [hunk 19]
```diff
@@ -36,17 +36,22 @@ export default defineComponent({
...
       try {
```
Comment (dheepak-aot) [inline sources/packages/web/src/views/aest/institution/ProfileCreate.vue:47]: same as https://github.com/bcgov/SIMS/pull/2045/files#r1240372556

**Comment by dheepak-aot** [comment]
Great job 👍Added some comments.

File: sources/packages/backend/apps/api/src/utilities/class-transform/to-boolean.ts [hunk 20]
```diff
@@ -0,0 +1,40 @@
+import { Transform } from "class-transformer";
+
+/**
+ * Converts a plain property to boolean.
+ * @returns converted boolean value.
+ */
+export const ToBoolean = () => {
```
Comment (andrepestana-aot) [inline sources/packages/backend/apps/api/src/utilities/class-transform/to-boolean.ts:7]: For my knowledge: should we add return type for these functions?

**Review by andrepestana-aot — COMMENTED** [review]
Great job! Added a question.

File: sources/packages/backend/apps/api/src/route-controllers/institution-locations/models/application.dto.ts (hunk 2 above)
Comment (andrewsignori-aot) [reply sources/packages/backend/apps/api/src/route-controllers/institution-locations/models/application.dto.ts:55]: They are nicely cast when DTOs ate from the body payload where a boolean is actually received as a boolean.
```ts
{
    // This will be cast to a boolean.
    myBolleanProperty: true
}
```
When the values come from the query string, all of them are strings, which will make any string different than the "empty string" `true`.

File: sources/packages/backend/apps/api/src/route-controllers/institution-locations/institution-location.institutions.controller.ts (hunk 3 above)
Comment (andrewsignori-aot) [reply sources/packages/backend/apps/api/src/route-controllers/institution-locations/institution-location.institutions.controller.ts:158]: I will update the comments,
The `queryStringDTO` was meant to be generic. It is supposed to represent any query string received by the endpoint, in this case, paginations options + archive flag.
I avoided using a name that represents "paginations options + archive flag" thinking that in the future if a new parameter is added we should not rename the parameter again.

File: sources/packages/backend/apps/api/src/route-controllers/institution-locations/models/application.dto.ts (hunk 8 above)
Comment (andrewsignori-aot) [reply sources/packages/backend/apps/api/src/route-controllers/institution-locations/models/application.dto.ts:53]: Up to us, I was trying to make some differences between the payloads from the body.
We can keep the `APIInDTO` if the team prefers.
@dheepak-aot @guru-aot @andrepestana-aot @sh16011993 thoughts?

File: sources/packages/web/src/services/http/dto/Common.dto.ts (hunk 11 above)
Comment (andrewsignori-aot) [reply sources/packages/web/src/services/http/dto/Common.dto.ts:21]: Please see the comments, they are from 14 months ago. I will not be changing it.
image

File: sources/packages/web/src/components/partial-view/DesignationAgreement/DesignationAgreementForm.models.ts (hunk 16 above)
Comment (andrewsignori-aot) [reply sources/packages/web/src/components/partial-view/DesignationAgreement/DesignationAgreementForm.models.ts:55]: Yes, they are. This model is used to map the UI, from this model the DTO will be created and there they should be required.

File: sources/packages/web/src/views/aest/institution/ProfileCreate.vue (hunk 17 above)
Comment (andrewsignori-aot) [reply sources/packages/web/src/views/aest/institution/ProfileCreate.vue:46]: We can keep it mapped as `AESTCreateInstitutionAPIInDTO` or whatever we want. Either way, the form.io will not have the type enforced. The idea of having it as `unknown` was to ensure that this `data` would not be used anywhere before it is properly typed.
If need access to some property not mapped to `AESTCreateInstitutionAPIInDTO` we probably should have another type mapped.
If it creates more confusion as `unknown` I can revert it.

File: sources/packages/backend/apps/api/src/utilities/class-transform/to-boolean.ts (hunk 20 above)
Comment (andrewsignori-aot) [reply sources/packages/backend/apps/api/src/utilities/class-transform/to-boolean.ts:7]: We can, just the return type will be a long repetition of the next line. If there is an understand that it would be clear in this way I can add it as shown below:
```ts
/**
 * Converts a plain property to boolean.
 * @returns converted boolean value.
 */
export const ToBoolean = (): ((target: any, key: string) => void) => {
  return (target: any, key: string) => {
    return Transform(
      ({ obj }) => {
        return valueToBoolean(obj[key]);
      },
      {
        toClassOnly: true,
      },
    )(target, key);
  };
};
```

File: sources/packages/backend/apps/api/src/services/education-program/education-program.service.models.ts (hunk 4 above)
Comment (andrewsignori-aot) [reply sources/packages/backend/apps/api/src/services/education-program/education-program.service.models.ts]: Changes reverted.

File: sources/packages/backend/apps/api/src/route-controllers/education-program/_tests_/e2e/education-program.institutions.controller.createEducationProgram.e2e-spec.ts (hunk 6 above)
Comment (andrewsignori-aot) [reply sources/packages/backend/apps/api/src/route-controllers/education-program/_tests_/e2e/education-program.institutions.controller.createEducationProgram.e2e-spec.ts]: Removed.

File: sources/packages/backend/apps/api/src/route-controllers/institution-locations/models/application.dto.ts (hunk 8 above)
Comment (andrewsignori-aot) [reply sources/packages/backend/apps/api/src/route-controllers/institution-locations/models/application.dto.ts:53]: Renamed to `ActiveApplicationSummaryQueryStringAPIInDTO` to keep `APIinDTO` suffix.

File: sources/packages/web/src/services/InstitutionService.ts (hunk 10 above)
Comment (andrewsignori-aot) [reply sources/packages/web/src/services/InstitutionService.ts:55]: Comments added.

File: sources/packages/web/src/services/http/dto/ScholasticStanding.dto.ts (hunk 12 above)
Comment (andrewsignori-aot) [reply sources/packages/web/src/services/http/dto/ScholasticStanding.dto.ts:7]: Only DTOs that will have their type enforced were changed to a class to have the class transformation possible.

File: sources/packages/web/src/services/http/dto/Institution.dto.ts (hunk 14 above)
Comment (andrewsignori-aot) [reply sources/packages/web/src/services/http/dto/Institution.dto.ts:127]: The properties from the interface `CreateInstitution` are implemented in the class `CreateInstitutionAPIInDTO` and properly decorated.
image

File: sources/packages/web/src/views/aest/institution/InstitutionLocationEdit.vue (hunk 18 above)
Comment (andrewsignori-aot) [reply sources/packages/web/src/views/aest/institution/InstitutionLocationEdit.vue:41]: At this moment I am not sure what would be clear for a developer in the future.
Maybe the ideal scenario would have a dedicated model for every form.io and then have it properly mapped.
This form.io, for instance, I hope that it would be converted to a regular Vuetify someday.
For now, I am changing it to unknow.

File: sources/packages/web/src/views/aest/institution/ProfileCreate.vue (hunk 19 above)
Comment (andrewsignori-aot) [reply sources/packages/web/src/views/aest/institution/ProfileCreate.vue:47]: Changed the liked comment to `unknow` and added a reply to it.

**Review by ann-aot — APPROVED** [review]
Thanks for doing the changes 👍 LGTM

File: sources/packages/backend/apps/api/src/route-controllers/institution-locations/models/application.dto.ts (hunk 2 above)
Comment (guru-aot) [reply sources/packages/backend/apps/api/src/route-controllers/institution-locations/models/application.dto.ts:55]: does this part of the wava scan suggestion?

File: sources/packages/web/src/composables/useFormioUtils.ts [hunk 21]
```diff
@@ -162,6 +163,20 @@ export function useFormioUtils() {
...
   };
 
+  /**
+   * Indicates if extraneous properties should be excluded from object.
+   * @param type class type to be used to identify the extraneous properties.
+   * @param plainObject object to be checked and have extraneous properties removed.
+   * @returns the plain object provided converted to the class type with no
+   * extraneous properties.
+   */
+  const excludeExtraneousValues = <T, V>(
```
Comment (guru-aot) [inline sources/packages/web/src/composables/useFormioUtils.ts:173]: 👍

File: sources/packages/web/src/services/http/dto/ApplicationException.dto.ts [hunk 22]
```diff
@@ -9,8 +10,10 @@ export interface CreateApplicationExceptionAPIInDTO {
   exceptionRequests: ApplicationExceptionRequestAPIInDTO[];
 }
 
-export interface UpdateApplicationExceptionAPIInDTO {
+export class UpdateApplicationExceptionAPIInDTO {
+  @Expose()
```
Comment (guru-aot) [inline sources/packages/web/src/services/http/dto/ApplicationException.dto.ts:14]: why we need this explicitly mentioned?

File: sources/packages/web/src/services/http/dto/Common.dto.ts (hunk 11 above)
Comment (dheepak-aot) [reply sources/packages/web/src/services/http/dto/Common.dto.ts:21]: I didn't notice that the name just came from an existing interface and I'm ok with the name being as it is.

File: sources/packages/web/src/services/http/dto/Common.dto.ts [hunk 23]
```diff
@@ -16,16 +18,26 @@ export interface InstitutionPrimaryContactAPIOutDTO {
...
+  @Expose()
   addressLine1: string;
+  @Expose()
```
Comment (guru-aot) [inline sources/packages/web/src/services/http/dto/Common.dto.ts:24]: Do we still need @Expose for optional parameters? Just for my understanding.

File: sources/packages/web/src/services/http/dto/Institution.dto.ts (hunk 14 above)
Comment (dheepak-aot) [reply sources/packages/web/src/services/http/dto/Institution.dto.ts:127]: If `CreateInstitution` is declared as a class and added `@expose` to it's properties we can just extend it right(without having to re-declare it's properties)?

Let me know if I am missing something.

File: sources/packages/backend/apps/api/src/route-controllers/institution-locations/models/application.dto.ts (hunk 2 above)
Comment (andrewsignori-aot) [reply sources/packages/backend/apps/api/src/route-controllers/institution-locations/models/application.dto.ts:55]: > does this part of the wava scan suggestion?

Just replying here as history since we had that clarify over a call, yes, this was needed otherwise the endpoint would fail because the `archive` was not defined in the DTO.

File: sources/packages/web/src/services/http/dto/ScholasticStanding.dto.ts (hunk 12 above)
Comment (dheepak-aot) [reply sources/packages/web/src/services/http/dto/ScholasticStanding.dto.ts:7]: Ah I see. I was under the impression that we will be starting to use class instead of interface for the DTOs in Vue side.

File: sources/packages/backend/apps/api/src/route-controllers/institution-locations/models/application.dto.ts (hunk 2 above)
Comment (dheepak-aot) [reply sources/packages/backend/apps/api/src/route-controllers/institution-locations/models/application.dto.ts:55]: Let me know if I am missing something. I get that values coming from query params are string and was trying to understand why we are not explicitly casting params like pageLimit and page which are numbers and only doing that for boolean field?

File: sources/packages/web/src/services/http/dto/Institution.dto.ts (hunk 14 above)
Comment (dheepak-aot) [reply sources/packages/web/src/services/http/dto/Institution.dto.ts]: Thanks for the change.

File: sources/packages/backend/apps/api/src/route-controllers/institution-locations/models/application.dto.ts (hunk 2 above)
Comment (andrewsignori-aot) [reply sources/packages/backend/apps/api/src/route-controllers/institution-locations/models/application.dto.ts:55]: > 

Maybe it is better to look on the below example:
image

**Review by guru-aot — APPROVED** [review]
LGTM, nice work @andrewsignori-aot

**Review by dheepak-aot — APPROVED** [review]
Thanks for making the changes 👍

File: sources/packages/web/src/services/http/dto/Common.dto.ts (hunk 23 above)
Comment (andrewsignori-aot) [reply sources/packages/web/src/services/http/dto/Common.dto.ts:24]: Just replying here as history since we had that clarify over a call, we need all properties expected to be part of the DTO decorated with `@Expose()` ([more info](https://github.com/typestack/class-transformer#enforcing-type-safe-instance)).

File: sources/packages/web/src/services/http/dto/ApplicationException.dto.ts (hunk 22 above)
Comment (andrewsignori-aot) [reply sources/packages/web/src/services/http/dto/ApplicationException.dto.ts:14]: Same from https://github.com/bcgov/SIMS/pull/2045#discussion_r1242859860

**Review by andrepestana-aot — APPROVED** [review]
Great job!
